src/WinRT.Runtime2/InteropServices/RestrictedErrorInfo.cs

```diff
@@ -62,26 +204,125 @@ public static void ThrowExceptionForHR(HRESULT errorCode)
+                    }
+                }
+                else if (isLanguageException)
+                {
+                    // Remove object reference to avoid cycles between error info holding exception
+                    // and exception holding error info.  We currently can't avoid this cycle
+                    // when the C# exception is caught on the C# side.
+                    exception.Data.Remove("__RestrictedErrorObjectReference");
+                    exception.Data.Remove("__HasRestrictedLanguageErrorObject");
+                }
+                if (restrictedErrorObject != null)
+                {
+                    using WindowsRuntimeObjectReferenceValue restrictedErrorObjectValue = restrictedErrorObject.AsValue();
+                    _ = WindowsRuntimeImports.SetRestrictedErrorInfo(restrictedErrorObjectValue.GetThisPtrUnsafe());
+                }
+            }
+            else
+            {
+                string message = exception.Message;
+                if (string.IsNullOrEmpty(message))
+                {
+                    Type exceptionType = exception.GetType();
+                    if (exceptionType != null)
+                    {
+                        message = exceptionType.Name;
+                    }
+                }
+
+                fixed (char* lpMessage = message)
+                {
+                    HStringMarshaller.ConvertToUnmanagedUnsafe(lpMessage, message.Length, out HStringReference hstring);
+                    WindowsRuntimeObjectReferenceValue managedExceptionWrapper = WindowsRuntimeObjectMarshaller.ConvertToUnmanaged(exception);
+                    try
+                    {
+                        _ = WindowsRuntimeImports.RoOriginateLanguageException(GetHRForException(exception), hstring.HString, managedExceptionWrapper.GetThisPtrUnsafe());
+                    }
+                    finally
+                    {
+                        _ = Marshal.Release((nint)managedExceptionWrapper.DetachThisPtrUnsafe());
+                    }
+
+                }
+            }
+        }
+        catch (Exception e)
+        {
+            // If we fail to set the error info, we continue on reporting the original exception.
+            Debug.Assert(false, e.Message, e.StackTrace);
+        }
     }
 
     /// <summary>
     /// Triggers the global error handler when an unhandled exception occurs.
     /// </summary>
     /// <param name="exception">The input <see cref="Exception"/> instance to flow to the global error handler.</param>
     /// <see href="https://learn.microsoft.com/windows/win32/api/roerrorapi/nf-roerrorapi-roreportunhandlederror"/>
-    public static void ReportUnhandledError(Exception exception)
+    public static unsafe void ReportUnhandledError(Exception exception)
     {
-        // TODO
+        SetErrorInfo(exception);
+        WindowsRuntimeObjectReferenceValue restrictedErrorInfoValue = ExceptionHelpers.BorrowRestrictedErrorInfo();
```

Comment:
Use a `using` block here instead of conditionally disposing. If the pointer is `null` the dispose will just be a no-op